File: src/recorder.js
const { createWriteStream } = require('fs');
const { pipeline } = require('stream/promises');
const path = require('path');
const { createAudioResource, createAudioPlayer, AudioPlayerStatus, EndBehaviorType } = require('@discordjs/voice');
const { OpusEncoder } = require('@discordjs/opus');
const prism = require('prism-media');
const { uploadToDrive } = require('./drive');
const config = require('../config.json');

/**
 * 音声チャンネルの録音を開始
 * @param {VoiceConnection} connection - 音声接続
 * @param {VoiceChannel} channel - 音声チャンネル
 * @param {Logger} logger - ロガー
 * @returns {Object} 録音セッション情報
 */
async function startRecording(connection, channel, logger) {
  try {
    // 録音ファイル名の生成（日時とチャンネル名を含む）
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const sanitizedChannelName = channel.name.replace(/[^\w-]/g, '_');
    const filename = `${timestamp}_${sanitizedChannelName}.${config.recordingFormat}`;
    const outputPath = path.join(config.localStoragePath, filename);
    
    logger.info(`録音ファイル作成: ${outputPath}`);
    
    // 録音ストリームの設定
    const receiver = connection.receiver;
    
    // 音声プレイヤーの作成（サイレントオーディオを再生して接続を維持）
    const player = createAudioPlayer();
    connection.subscribe(player);
    
    // 録音ストリームとファイル出力の設定
    const opusEncoder = new OpusEncoder(48000, 2);
    const ffmpegArgs = [
      '-f', 's16le',
      '-ar', '48000',
      '-ac', '2',
      '-c:a', 'libmp3lame',
      '-b:a', `${config.recordingBitrate}k`,
      outputPath
    ];
    
    // FFmpegプロセスの作成
    const ffmpeg = new prism.FFmpeg({
      args: ffmpegArgs
    });
    
    // 出力ファイルストリーム
    const outputStream = createWriteStream(outputPath);
    
    // 全ユーザーの音声を混合するためのマップ
    const userAudioStreams = new Map();
    
    // ユーザーが話し始めたときのイベント
    receiver.speaking.on('start', (userId) => {
      if (userAudioStreams.has(userId)) return;
      
      logger.debug(`ユーザー音声検出: ${userId}`);
      
      // ユーザーの音声ストリームを取得
      const audioStream = receiver.subscribe(userId, {
        end: {
          behavior: EndBehaviorType.AfterSilence,
          duration: 500
        }
      });
      
      // ストリームをFFmpegに接続
      audioStream.pipe(ffmpeg, { end: false });
      
      // ユーザーストリームを追跡
      userAudioStreams.set(userId, audioStream);
      
      // ストリーム終了時の処理
      audioStream.on('end', () => {
        logger.debug(`ユーザー音声終了: ${userId}`);
        userAudioStreams.delete(userId);
      });
    });
    
    // FFmpegの出力を録音ファイルに書き込む
    ffmpeg.pipe(outputStream, { end: false });
    
    // 録音セッション情報を返す
    return {
      connection,
      receiver,
      outputPath,
      filename,
      ffmpeg,
      outputStream,
      userAudioStreams,
      startTime: new Date()
    };
  } catch (error) {
    logger.error(`録音開始中にエラーが発生しました: ${error.message}`);
    logger.error(error.stack);
    throw error;
  }
}

/**
 * 録音を停止し、Google Driveにアップロード
 * @param {Object} recordingSession - 録音セッション情報
 * @param {VoiceChannel} channel - 音声チャンネル
 * @param {Logger} logger - ロガー
 * @returns {Object} 録音結果（Google Driveリンクなど）
 */
async function stopRecording(recordingSession, channel, logger) {
  try {
    const { outputPath, filename, ffmpeg, outputStream, userAudioStreams, startTime } = recordingSession;
    
    // 録音時間を計算
    const duration = (new Date() - startTime) / 1000; // 秒単位
    
    // 録音時間が短すぎる場合はスキップ（ノイズや誤検出の可能性）
    if (duration < 3) {
      logger.info(`録音時間が短すぎるためスキップします: ${duration}秒`);
      return null;
    }
    
    logger.info(`録音停止: ${outputPath} (${duration}秒)`);
    
    // すべてのユーザーストリームを終了
    for (const [userId, stream] of userAudioStreams.entries()) {
      stream.destroy();
    }
    
    // FFmpegストリームを終了
    ffmpeg.end();
    
    // 出力ストリームを閉じる
    outputStream.end();
    
    // ストリームが完全に閉じるのを待つ
    await new Promise(resolve => setTimeout(resolve, 1000));
    
    // Google Driveにアップロード
    logger.info(`Google Driveにアップロード中: ${filename}`);
    
    try {
      const driveResult = await uploadToDrive(outputPath, filename, channel.name, logger);
      
      // 自動削除が有効な場合、ローカルファイルを削除
      if (config.autoDeleteLocalFiles) {
        const fs = require('fs');
        fs.unlinkSync(outputPath);
        logger.info(`ローカルファイルを削除しました: ${outputPath}`);
      }
      
      return {
        filename,
        duration,
        driveLink: driveResult.webViewLink,
        driveId: driveResult.id
      };
    } catch (uploadError) {
      logger.error(`Google Driveへのアップロード中にエラーが発生しました: ${uploadError.message}`);
      logger.error(uploadError.stack);
      
      // アップロードに失敗してもローカルファイルは保持
      return {
        filename,
        duration,
        localPath: outputPath,
        error: uploadError.message
      };
    }
  } catch (error) {
    logger.error(`録音停止中にエラーが発生しました: ${error.message}`);
    logger.error(error.stack);
    return null;
  }
}

module.exports = {
  startRecording,
  stopRecording
};

File: python/src/recorder.py
import os
import asyncio
import datetime
from pathlib import Path
import discord
from discord.sinks import MP3Sink, WaveSink, OGGSink
from .drive import upload_to_drive

def get_sink_class(format_name):
    """
    録音フォーマットに基づいてSinkクラスを取得
    
    Args:
        format_name (str): 録音フォーマット名
        
    Returns:
        discord.sinks.Sink: 適切なSinkクラス
    """
    format_map = {
        'mp3': MP3Sink,
        'wav': WaveSink,
        'ogg': OGGSink
    }
    return format_map.get(format_name.lower(), MP3Sink)

async def start_recording(voice_client, channel, logger):
    """
    音声チャンネルの録音を開始
    
    Args:
        voice_client (discord.VoiceClient): 音声クライアント
        channel (discord.VoiceChannel): 音声チャンネル
        logger (Logger): ロガー
        
    Returns:
        dict: 録音セッション情報
    """
    try:
        # 設定ファイルの読み込み
        import json
        with open('config.json', 'r', encoding='utf-8') as f:
            config = json.load(f)
        
        # 録音ファイル名の生成（日時とチャンネル名を含む）
        timestamp = datetime.datetime.now().strftime('%Y-%m-%dT%H-%M-%S')
        sanitized_channel_name = ''.join(c if c.isalnum() or c in '-_' else '_' for c in channel.name)
        filename = f"{timestamp}_{sanitized_channel_name}.{config['recordingFormat']}"
        output_path = os.path.join(config['localStoragePath'], filename)
        
        logger.info(f'録音ファイル作成: {output_path}')
        
        # 録音フォーマットに基づいてSinkを選択
        sink_class = get_sink_class(config['recordingFormat'])
        
        # 録音の開始
        sink = sink_class(filters={"bitrate": config.get('recordingBitrate', 128) * 1000})
        voice_client.start_recording(
            sink,
            finished_callback=lambda sink: logger.debug('録音コールバック: 録音完了'),
            failed_callback=lambda err: logger.error(f'録音エラー: {err}')
        )
        
        # 録音セッション情報を返す
        return {
            'voice_client': voice_client,
            'output_path': output_path,
            'filename': filename,
            'sink': sink,
            'start_time': datetime.datetime.now()
        }
    except Exception as e:
        logger.error(f'録音開始中にエラーが発生しました: {str(e)}', exc_info=True)
        raise

async def stop_recording(recording_session, channel, logger):
    """
    録音を停止し、Google Driveにアップロード
    
    Args:
        recording_session (dict): 録音セッション情報
        channel (discord.VoiceChannel): 音声チャンネル
        logger (Logger): ロガー
        
    Returns:
        dict: 録音結果（Google Driveリンクなど）
    """
    try:
        voice_client = recording_session['voice_client']
        output_path = recording_session['output_path']
        filename = recording_session['filename']
        sink = recording_session['sink']
        start_time = recording_session['start_time']
        
        # 録音時間を計算
        duration = (datetime.datetime.now() - start_time).total_seconds()
        
        # 録音時間が短すぎる場合はスキップ（ノイズや誤検出の可能性）
        if duration < 3:
            logger.info(f'録音時間が短すぎるためスキップします: {duration}秒')
            return None
        
        logger.info(f'録音停止: {output_path} ({duration}秒)')
        
        # 録音を停止
        voice_client.stop_recording()
        
        # 録音データを保存
        os.makedirs(os.path.dirname(output_path), exist_ok=True)
        
        # 設定ファイルの読み込み
        import json
        with open('config.json', 'r', encoding='utf-8') as f:
            config = json.load(f)
        
        # 録音データをファイルに書き込む
        audio_data = sink.get_all_audio()
        if not audio_data:
            logger.warning('録音データがありません')
            return None
        
        # 複数ユーザーの音声を1つのファイルに結合
        # 注: 実際の実装ではより複雑な音声ミキシングが必要かもしれません
        with open(output_path, 'wb') as f:
            # 最初のユーザーの音声データを使用
            for user_id, audio in audio_data.items():
                f.write(audio.file.read())
                break
        
        # Google Driveにアップロード
        logger.info(f'Google Driveにアップロード中: {filename}')
        
        try:
            drive_result = await upload_to_drive(output_path, filename, channel.name, logger)
            
            # 自動削除が有効な場合、ローカルファイルを削除
            if config.get('autoDeleteLocalFiles', True):
                os.unlink(output_path)
                logger.info(f'ローカルファイルを削除しました: {output_path}')
            
            return {
                'filename': filename,
                'duration': duration,
                'drive_link': drive_result.get('web_view_link'),
                'drive_id': drive_result.get('id')
            }
        except Exception as upload_error:
            logger.error(f'Google Driveへのアップロード中にエラーが発生しました: {str(upload_error)}', exc_info=True)
            
            # アップロードに失敗してもローカルファイルは保持
            return {
                'filename': filename,
                'duration': duration,
                'local_path': output_path,
                'error': str(upload_error)
            }
    except Exception as e:
        logger.error(f'録音停止中にエラーが発生しました: {str(e)}', exc_info=True)
        return None